### PR TITLE
Unit Test code for repo and task: #1

### PR DESCRIPTION
**Pull request by oliveiradan**
Unity test for repo_test.go and task_test.go (1st attempt).

### CODE REVIEW

File: datastore/bolt/repo_test.go [hunk 1]
```diff
@@ -1,24 +1,132 @@
...
+	})
+
+	//--
+	Convey("Should delete", t, func() {
+		err := db.DeleteRepo(testRepo)
+		So(err, ShouldBeNil)
+	})
+
+	//-- Delete the temp db at the end.
+	os.Remove(db.Path())
```
Comment (bradrydzewski) [inline datastore/bolt/repo_test.go]: use the Go defer statement and put these two together
```
var db = Must("/tmp/drone.test.db")
defer os.Remove(db.Path())
```

File: datastore/bolt/repo_test.go [hunk 2]
```diff
@@ -1,24 +1,132 @@
...
+		g.It("Should delete") // ok
+		g.It("Should insert") // ok
+		g.It("Should not insert if exists")  // ?
 		g.It("Should insert params")
 		g.It("Should update params")
 		g.It("Should insert keys")
 		g.It("Should update keys")
 	})
+	*/
+	var testUser string = "octocat"
```
Comment (bradrydzewski) [inline datastore/bolt/repo_test.go]: no need to declare types here. Go infers them automatically. Using `var` is also unnecessary in this case:
``` 
testUser := "octocat"
```

File: datastore/bolt/repo_test.go [hunk 3]
```diff
@@ -1,24 +1,132 @@
...
+	testParamsUpd := map[string]string{
+		"A": "Alpha-Upd",
+		"B": "Beta-Upd",
+		"C": "Charlie-Upd",
+		"D": "Delta-Upd",
+		"E": "Echo-Upd",
+	}
+
+	//--
+	db = Must("/tmp/drone.test.db")
```
Comment (bradrydzewski) [inline datastore/bolt/repo_test.go]: in the other unit tests we follow this pattern:

```
		// create a new database before each unit
		// test and destroy afterwards.
		g.BeforeEach(func() {
			db = Must("/tmp/drone.test.db")
		})
		g.AfterEach(func() {
			os.Remove(db.Path())
		})
```

this creates and destroys a database for each unit test, which provides better isolation and prevents unit tests from conflicting with eachother

File: datastore/bolt/repo_test.go [hunk 4]
```diff
@@ -1,24 +1,132 @@
...
+		//-- generate an RSA key and add to the repo
+		testKey, err := sshutil.GeneratePrivateKey()
+		So(err, ShouldBeNil)
+		//--
+		testKeypair := &common.Keypair{}
+		testKeypair.Public = sshutil.MarshalPublicKey(&testKey.PublicKey)
+		testKeypair.Private = sshutil.MarshalPrivateKey(testKey)
+		//--
+		err := db.UpsertRepoKeys(testRepo, testKeypair)
+		So(err, ShouldBeNil)
```
Comment (bradrydzewski) [inline datastore/bolt/repo_test.go]: looks similar to "should insert keys". duplicate?

File: datastore/bolt/repo_test.go [hunk 5]
```diff
@@ -1,24 +1,132 @@
...
+	//--
+	Convey("Should insert params", t, func() {
+		err := db.UpsertRepoParams(testRepo, testParamsIns)
+		So(err, ShouldBeNil)
+	})
+
+	//--
+	Convey("Should update params", t, func() {
+		err := db.UpsertRepoParams(testRepo, testParamsUpd)
+		So(err, ShouldBeNil)
```
Comment (bradrydzewski) [inline datastore/bolt/repo_test.go]: should test the ability to retrieve the update params, and verify they were correctly persisted. See the `user_test.go` file for examples

File: datastore/bolt/repo_test.go [hunk 6]
```diff
@@ -1,24 +1,132 @@
...
+		"B": "Beta-Upd",
+		"C": "Charlie-Upd",
+		"D": "Delta-Upd",
+		"E": "Echo-Upd",
+	}
+
+	//--
+	db = Must("/tmp/drone.test.db")
+
+	//--
```
Comment (bradrydzewski) [inline datastore/bolt/repo_test.go]: remove `//--`

File: datastore/bolt/repo_test.go [hunk 7]
```diff
@@ -1,24 +1,132 @@
...
+	Convey("Should insert", t, func() {
+		err := db.InsertRepo(testUser, testRepo)
+		So(err, ShouldBeNil)
+		//So(InsertRepo(testUser, testRepo), ShouldE)
```
Comment (bradrydzewski) [inline datastore/bolt/repo_test.go]: should get the repo to verify it was correctly inserted. See `user_test.go` for reference

File: datastore/bolt/repo_test.go [hunk 8]
```diff
@@ -1,24 +1,132 @@
...
+	//--
+	Convey("Should find params", t, func() {
+		params, err := db.GetRepoParams(testRepo)
+		So(params, ShouldNotBeEmpty)
+		So(err, ShouldBeNil)
+	})
+
+	//--
+	Convey("Should find keys", t, func() {
+		keypair, err := db.GetRepoKeys(testRepo)
```
Comment (bradrydzewski) [inline datastore/bolt/repo_test.go]: should assert keys (public, private) not empty and have expected values. Also, where is it inserting these keys?

File: datastore/bolt/repo_test.go [hunk 9]
```diff
@@ -1,24 +1,132 @@
...
+
+	//-- Lets try to add the same repo again.
+	Convey("Should not insert if exists", t, func() {
+		err := db.InsertRepo(testUser, testRepo)
+		So(err, ShouldEqual, ErrKeyExists)
+	})
+
+	//--
+	Convey("Should find params", t, func() {
+		params, err := db.GetRepoParams(testRepo)
```
Comment (bradrydzewski) [inline datastore/bolt/repo_test.go]: should assert param map not empty and has expected values. Also, where is it inserting these params?

**Comment by bradrydzewski** [comment]
@oliveiradan are you running `go test` locally in the `datastore/bolt` directory? I see compiler errors:

```
./repo_test.go:5: imported and not used: "github.com/drone/drone/common/sshutil"
./repo_test.go:38: undefined: os
./repo_test.go:42: cannot use testRepo (type string) as type *common.Repo in argument to db.SetRepo
./repo_test.go:48: undefined: tesRepo
./repo_test.go:54: g.Assert(repos).NotEqual undefined (type *goblin.Assertion has no field or method NotEqual)
./repo_test.go:84: cannot use testUser (type string) as type *common.User in argument to db.SetRepoNotExists
./repo_test.go:84: cannot use testRepo (type string) as type *common.Repo in argument to db.SetRepoNotExists
./repo_test.go:85: g.Assert(err).NotEqual undefined (type *goblin.Assertion has no field or method NotEqual)
./repo_test.go:86: undefined: ErrConflict
./repo_test.go:90: cannot use testUser (type string) as type *common.User in argument to db.SetRepoNotExists
./repo_test.go:90: too many errors
FAIL	github.com/drone/drone/datastore/bolt [build failed]
```

Also, I'm not sure the tests will pass as written. For example the below test attempts to get a repository, however, a repository has not been inserted into the database.

```
		g.It("Should get Repo", func() {
			repo, err := db.Repo(testRepo)
			g.Assert(repo).Equal(tesRepo)
			g.Assert(err).Equal(nil)
		})
```

Notice in the `user_test.go` file we insert the user first, and then we get the user, and then we verify the fields match what were inserted:

```
		g.It("Should find", func() {
			db.SetUserNotExists(&common.User{Login: "octocat"})
			user, err := db.User("octocat")
			g.Assert(err).Equal(nil)
			g.Assert(user.Login).Equal("octocat")
		})
```

Also, the PR still includes files from master. It looks like you pulled down some recent changes from master. If you see the changelog in this pull requests, there are 6 commits that came from master:

image

Those commits were only intended for master and should not be merged into the bolt branch. These files should therefore not be pushed:

* shared/build/buildfile/buildfile_test.go
* shared/build/buildfile/buildfile.go
* remote/github/helper.go

**Comment by oliveiradan** [comment]
@bradrydzewski Thanks for taking the time to help fixing my dev. environment. 
I think we are in a better shape now and both of the unity tests were rewritten.

doliveira@doliveira-opensuse-dev10:~/go/src/github.com/drone/drone> clear; date; make test
Mon Apr 20 15:19:14 MDT 2015
go vet ./...
go test -cover -short ./...
?       github.com/drone/drone  [no test files]
?       github.com/drone/drone/builder  [no test files]
?       github.com/drone/drone/builder/docker   [no test files]
?       github.com/drone/drone/builder/pool     [no test files]
?       github.com/drone/drone/common   [no test files]
?       github.com/drone/drone/common/ccmenu    [no test files]
?       github.com/drone/drone/common/httputil  [no test files]
?       github.com/drone/drone/common/oauth2    [no test files]
?       github.com/drone/drone/common/sshutil   [no test files]
?       github.com/drone/drone/datastore        [no test files]
ok      github.com/drone/drone/datastore/bolt   10.966s coverage: 72.6% of statements
?       github.com/drone/drone/datastore/rpc    [no test files]
?       github.com/drone/drone/eventbus [no test files]
ok      github.com/drone/drone/parser   0.002s  coverage: 67.5% of statements
ok      github.com/drone/drone/parser/inject    0.002s  coverage: 100.0% of statements
ok      github.com/drone/drone/parser/matrix    0.002s  coverage: 68.6% of statements
?       github.com/drone/drone/remote   [no test files]
?       github.com/drone/drone/remote/client    [no test files]
?       github.com/drone/drone/remote/github    [no test files]
?       github.com/drone/drone/server   [no test files]
?       github.com/drone/drone/server/session   [no test files]
?       github.com/drone/drone/settings [no test files]
doliveira@doliveira-opensuse-dev10:~/go/src/github.com/drone/drone>

**Comment by bradrydzewski** [comment]
cool. Let's close this pull request (since it was from your old fork) and then you can re-open a fresh pull request. Sound good?

**Comment by oliveiradan** [comment]
yep. Thanks again!

**Pull request closed without merging**